File: repos/rappterbook-agent/python/openrappter/agents/broadcast.py
"""
BroadcastManager - Send messages to multiple agents simultaneously.

Modes:
  all      - Send to all agents, wait for all responses
  race     - Send to all agents, return first success
  fallback - Try agents sequentially until one succeeds

In fallback mode, data_slush from each failed agent is passed as
upstream_slush to the next agent in line.

Mirrors TypeScript agents/broadcast.ts
"""

import asyncio


class BroadcastManager:
    """Manages broadcast groups for multi-agent messaging."""

    def __init__(self):
        self._groups = {}

    def create_group(self, group):
        """Create a broadcast group.

        Args:
            group: dict with keys: id, name, agentIds, mode, timeout (optional)
        """
        self._groups[group['id']] = group

    def remove_group(self, group_id):
        """Remove a broadcast group. Returns True if removed."""
        if group_id in self._groups:
            del self._groups[group_id]
            return True
        return False

    def get_group(self, group_id):
        """Get a broadcast group by ID."""
        return self._groups.get(group_id)

    def get_groups(self):
        """Get all broadcast groups."""
        return list(self._groups.values())

    async def broadcast(self, group_id, message, executor):
        """Broadcast message to a group.

        Args:
            group_id: The broadcast group ID
            message: The message to send
            executor: async callable(agent_id, message, upstream_slush=None) -> AgentResult

        Returns:
            BroadcastResult dict
        """
        group = self._groups.get(group_id)
        if not group:
            raise ValueError(f"Broadcast group not found: {group_id}")

        mode = group.get('mode', 'all')
        if mode == 'all':
            return await self._broadcast_all(group, message, executor)
        elif mode == 'race':
            return await self._broadcast_race(group, message, executor)
        elif mode == 'fallback':
            return await self._broadcast_fallback(group, message, executor)
        else:
            return await self._broadcast_all(group, message, executor)

    async def _broadcast_all(self, group, message, executor):
        """Broadcast to all agents and wait for all responses."""
        results = {}
        first_response = None
        timeout = group.get('timeout')

        async def run_agent(agent_id):
            nonlocal first_response
            try:
                coro = executor(agent_id, message)
                if timeout:
                    result = await asyncio.wait_for(coro, timeout=timeout / 1000)
                else:
                    result = await coro
                results[agent_id] = result
                if first_response is None:
                    first_response = {'agentId': agent_id, 'result': result}
            except Exception as e:
                results[agent_id] = e

        await asyncio.gather(*[run_agent(aid) for aid in group['agentIds']], return_exceptions=True)

        successes = sum(1 for r in results.values() if not isinstance(r, Exception))

        return {
            'groupId': group['id'],
            'results': results,
            'firstResponse': first_response,
            'allSucceeded': successes == len(group['agentIds']),
            'anySucceeded': successes > 0,
        }

    async def _broadcast_race(self, group, message, executor):
        """Broadcast to all agents and return first response."""
        results = {}
        first_response = None
        timeout = group.get('timeout')

        async def run_agent(agent_id):
            try:
                coro = executor(agent_id, message)
                if timeout:
                    result = await asyncio.wait_for(coro, timeout=timeout / 1000)
                else:
                    result = await coro
                results[agent_id] = result
                return {'agentId': agent_id, 'result': result, 'success': True}
            except Exception as e:
                results[agent_id] = e
                raise

        tasks = [asyncio.create_task(run_agent(aid)) for aid in group['agentIds']]

        # Wait for first successful result
        try:
            done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)

            for task in done:
                if not task.cancelled() and task.exception() is None:
                    r = task.result()
                    if r.get('success'):
                        first_response = {'agentId': r['agentId'], 'result': r['result']}
                        break

            # Cancel pending tasks
            for task in pending:
                task.cancel()

            # Wait for remaining to finish
            if pending:
                await asyncio.gather(*pending, return_exceptions=True)
        except Exception:
            pass

        # Wait for all tasks to be done
        await asyncio.gather(*tasks, return_exceptions=True)

        successes = sum(1 for r in results.values() if not isinstance(r, Exception))

        return {
            'groupId': group['id'],
            'results': results,
            'firstResponse': first_response,
            'allSucceeded': successes == len(group['agentIds']),
            'anySucceeded': successes > 0,
        }

    async def _broadcast_fallback(self, group, message, executor):
        """Try agents in order until one succeeds.

        In fallback mode, data_slush from each failed agent is passed
        as upstream_slush to the next agent.
        """
        results = {}
        first_response = None
        last_slush = None
        timeout = group.get('timeout')

        for agent_id in group['agentIds']:
            try:
                coro = executor(agent_id, message, last_slush)
                if timeout:
                    result = await asyncio.wait_for(coro, timeout=timeout / 1000)
                else:
                    result = await coro
                results[agent_id] = result
                first_response = {'agentId': agent_id, 'result': result}
                break  # Success, stop trying
            except Exception as e:
                results[agent_id] = e
                # Extract data_slush from failed agent for downstream chaining
                if hasattr(e, 'result') and isinstance(e.result, dict) and 'data_slush' in e.result:
                    last_slush = e.result['data_slush']

        successes = sum(1 for r in results.values() if not isinstance(r, Exception))

        return {
            'groupId': group['id'],
            'results': results,
            'firstResponse': first_response,
            'allSucceeded': successes == len(group['agentIds']),
            'anySucceeded': successes > 0,
        }


def create_broadcast_manager():
    """Factory function to create a BroadcastManager."""
    return BroadcastManager()

File: repos/rappterbook-agent/typescript/src/agents/broadcast.ts
/**
 * Broadcast Groups
 * Send messages to multiple agents simultaneously
 */

import type { AgentResult } from './types.js';

export interface BroadcastGroup {
  id: string;
  name: string;
  agentIds: string[];
  mode: 'all' | 'race' | 'fallback';
  timeout?: number;
}

export interface BroadcastResult {
  groupId: string;
  results: Map<string, AgentResult | Error>;
  firstResponse?: { agentId: string; result: AgentResult };
  allSucceeded: boolean;
  anySucceeded: boolean;
}

type AgentExecutor = (agentId: string, message: string, upstreamSlush?: Record<string, unknown>) => Promise<AgentResult>;

export class BroadcastManager {
  private groups = new Map<string, BroadcastGroup>();

  /**
   * Create a broadcast group
   */
  createGroup(group: BroadcastGroup): void {
    this.groups.set(group.id, group);
  }

  /**
   * Remove a broadcast group
   */
  removeGroup(groupId: string): boolean {
    return this.groups.delete(groupId);
  }

  /**
   * Get a broadcast group
   */
  getGroup(groupId: string): BroadcastGroup | undefined {
    return this.groups.get(groupId);
  }

  /**
   * Get all groups
   */
  getGroups(): BroadcastGroup[] {
    return Array.from(this.groups.values());
  }

  /**
   * Broadcast message to a group
   */
  async broadcast(
    groupId: string,
    message: string,
    executor: AgentExecutor
  ): Promise<BroadcastResult> {
    const group = this.groups.get(groupId);
    if (!group) {
      throw new Error(`Broadcast group not found: ${groupId}`);
    }

    switch (group.mode) {
      case 'all':
        return this.broadcastAll(group, message, executor);
      case 'race':
        return this.broadcastRace(group, message, executor);
      case 'fallback':
        return this.broadcastFallback(group, message, executor);
      default:
        return this.broadcastAll(group, message, executor);
    }
  }

  /**
   * Broadcast to all agents and wait for all responses
   */
  private async broadcastAll(
    group: BroadcastGroup,
    message: string,
    executor: AgentExecutor
  ): Promise<BroadcastResult> {
    const results = new Map<string, AgentResult | Error>();
    let firstResponse: { agentId: string; result: AgentResult } | undefined;

    const promises = group.agentIds.map(async (agentId) => {
      try {
        const result = await this.executeWithTimeout(executor, agentId, message, group.timeout);
        results.set(agentId, result);
        if (!firstResponse) {
          firstResponse = { agentId, result };
        }
        return { agentId, result, success: true };
      } catch (error) {
        results.set(agentId, error as Error);
        return { agentId, error, success: false };
      }
    });

    await Promise.all(promises);

    const successes = Array.from(results.values()).filter(
      (r) => !(r instanceof Error)
    ).length;

    return {
      groupId: group.id,
      results,
      firstResponse,
      allSucceeded: successes === group.agentIds.length,
      anySucceeded: successes > 0,
    };
  }

  /**
   * Broadcast to all agents and return first response
   */
  private async broadcastRace(
    group: BroadcastGroup,
    message: string,
    executor: AgentExecutor
  ): Promise<BroadcastResult> {
    const results = new Map<string, AgentResult | Error>();
    let firstResponse: { agentId: string; result: AgentResult } | undefined;

    const promises = group.agentIds.map(async (agentId) => {
      try {
        const result = await this.executeWithTimeout(executor, agentId, message, group.timeout);
        results.set(agentId, result);
        return { agentId, result, success: true };
      } catch (error) {
        results.set(agentId, error as Error);
        return { agentId, error, success: false };
      }
    });

    // Wait for first successful result
    try {
      const first = await Promise.race(
        promises.map((p) =>
          p.then((r) => {
            if (r.success) return r;
            throw r.error;
          })
        )
      );

      if (first.success && 'result' in first) {
        firstResponse = { agentId: first.agentId, result: first.result as AgentResult };
      }
    } catch {
      // All failed
    }

    // Wait for all to complete (for result map)
    await Promise.allSettled(promises);

    const successes = Array.from(results.values()).filter(
      (r) => !(r instanceof Error)
    ).length;

    return {
      groupId: group.id,
      results,
      firstResponse,
      allSucceeded: successes === group.agentIds.length,
      anySucceeded: successes > 0,
    };
  }

  /**
   * Try agents in order until one succeeds.
   * In fallback mode, data_slush from each failed agent is passed to the next.
   */
  private async broadcastFallback(
    group: BroadcastGroup,
    message: string,
    executor: AgentExecutor
  ): Promise<BroadcastResult> {
    const results = new Map<string, AgentResult | Error>();
    let firstResponse: { agentId: string; result: AgentResult } | undefined;
    let lastSlush: Record<string, unknown> | undefined;

    for (const agentId of group.agentIds) {
      try {
        const result = await this.executeWithTimeout(executor, agentId, message, group.timeout, lastSlush);
        results.set(agentId, result);
        firstResponse = { agentId, result };
        break; // Success, stop trying
      } catch (error) {
        results.set(agentId, error as Error);
        // Extract data_slush from failed agent for downstream chaining
        const err = error as Error & { result?: AgentResult };
        if (err.result && typeof err.result === 'object' && 'data_slush' in err.result) {
          lastSlush = err.result.data_slush as Record<string, unknown>;
        }
        // Continue to next agent
      }
    }

    const successes = Array.from(results.values()).filter(
      (r) => !(r instanceof Error)
    ).length;

    return {
      groupId: group.id,
      results,
      firstResponse,
      allSucceeded: successes === group.agentIds.length,
      anySucceeded: successes > 0,
    };
  }

  /**
   * Execute with optional timeout
   */
  private async executeWithTimeout(
    executor: AgentExecutor,
    agentId: string,
    message: string,
    timeout?: number,
    upstreamSlush?: Record<string, unknown>
  ): Promise<AgentResult> {
    if (!timeout) {
      return executor(agentId, message, upstreamSlush);
    }

    return Promise.race([
      executor(agentId, message, upstreamSlush),
      new Promise<AgentResult>((_, reject) =>
        setTimeout(() => reject(new Error(`Agent ${agentId} timeout after ${timeout}ms`)), timeout)
      ),
    ]);
  }
}

export function createBroadcastManager(): BroadcastManager {
  return new BroadcastManager();
}
